core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala

```diff
@@ -268,23 +243,11 @@ private[spark] class ListenerTaskMetrics(
 }
 
 private[spark] object TaskMetrics extends Logging {
+  import InternalAccumulator._
 
   def empty: TaskMetrics = new TaskMetrics
 
-  /**
-   * Get an accumulator from the given map by name, assuming it exists.
-   */
-  def getAccum[T](accumMap: Map[String, Accumulator[_]], name: String): Accumulator[T] = {
-    require(accumMap.contains(name), s"metric '$name' is missing")
-    val accum = accumMap(name)
-    try {
-      // Note: we can't do pattern matching here because types are erased by compile time
-      accum.asInstanceOf[Accumulator[T]]
-    } catch {
-      case e: ClassCastException =>
-        throw new SparkException(s"accumulator $name was of unexpected type", e)
-    }
-  }
+  def createAccum[T](name: String): Accumulator[T] = create(name).asInstanceOf[Accumulator[T]]
```

Comment:
I'd move the creation of accumulators in here, rather than delegating to InternalAccumulators.

Also maybe just have createLongAccumulator and createCollectionAccumulator; then it becomes obvious at the callsite what's going on, and we also don't need to have conditional branches in create.